Feat: pipelinerun reconciliation

hasanawad94 opened the pull request:
This will be relevant to https://github.com/shipwright-io/community/pull/277.
This pr only covers introducing `PipelineRuns` as an alternative to `taskruns` to be used for buildruns.
With these changes when `BUILDRUN_EXECUTOR` controller environment flag is set to `PipelineRun` the original `taskrun` will be embedded into a pipelinerun.

Future prs will cover making `PipelineRun` to define a set of `TaskRuns` that can run in parallel. This [shp](https://github.com/shipwright-io/community/pull/277) covers the details.

# Changes

- Add `pipelinerun_runner.go` with `TektonPipelineRunWrapper` implementation
  that wraps Tekton PipelineRuns and injects TaskRun specs into PipelineRun
  structures.
- Implement `ImageBuildRunner` and `ImageBuildRunnerFactory` interfaces for
  PipelineRun support.
- Add reconciliation logic for PipelineRun status updates, including
  condition handling and failure detection while maintaining original reconcile logic for `taskruns`.
- Add `UpdateBuildRunUsingPipelineRunCondition` function to update BuildRun
  status based on PipelineRun conditions (timeout, cancellation, success,
  failure)
- Refactor volume checking to use `GetUnderlyingTaskRun()` method for both
  TaskRun and PipelineRun executors, eliminating the need for separate
  volume checking methods
- Add `RunnerFactories` map to support configurable executor selection
  between TaskRun and PipelineRun implementations
- Update error handling in `CreateImageBuildRunner`
- Added unit, integration and e2e tests to ci.


# Submitter Checklist

- [x] Includes tests if functionality changed/was added
- [x] Includes docs if changes are user-facing
- [x] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)
- [x] Release notes block has been filled in, or marked NONE


# Release Notes

```release-note
Added support for Tekton PipelineRuns as an alternative to TaskRuns for build execution. Configurable executor selection - users can now choose between TaskRun and PipelineRun executors via configuration.
```

ayushsatyam146 commented on pkg/reconciler/buildrun/pipelinerun_runner.go:
```diff
@@ -0,0 +1,229 @@
+	// Since we only have one task for now in our PipelineRun, we can use the first ChildReference.
+	// The pod name will be the same as the TaskRun name with a suffix.
+	firstTaskRef := t.PipelineRun.Status.ChildReferences[0]
+	return firstTaskRef.Name + "-pod"
```
won't it keep creating the same pod name for the same taskRef even on different interations? I think it should be "-pod-" for unique name gen

hasanawad94 replied:
Yes so for this one TekTon adds a `-pod` suffix to the pod that gets generated from the taskrun. I thought I'd have it hard coded to not have to make an api call to fetch the podname

SaschaSchwarze0 replied:
Can the TaskRun name be too long so that Tekton does not append `-pod` but applies some additional magic to shorten it ? In general relying on an implementation detail of Tekton feels incorrect.

And general question: why do we return the name of the first Pod here ? In which context are we using this function (I guess maybe error reporting) and would it make sense to revisit those to see which pod name we should return here.

ayushsatyam146 commented on pkg/reconciler/buildrun/resources/pipelinerun.go:
```diff
@@ -0,0 +1,61 @@
+	// Generate a TaskRun object using the existing logic
+	taskRun, err := GenerateTaskRun(cfg, build, buildRun, serviceAccountName, strategy)
+	if err != nil {
+		return nil, err
```
maybe as descriptive error here would be helpful

ayushsatyam146 commented on pkg/reconciler/buildrun/imagebuildrunner.go:
```diff
@@ -184,6 +196,16 @@ func (t *TektonTaskRunWrapper) GetObject() client.Object {
 
+// GetExecutorKind returns the kind of executor.
+func (t *TektonTaskRunWrapper) GetExecutorKind() string {
+	return "TaskRun"
```
I am not aware of the full scale of the change, but would returning a string here be okay? Should we consider returning Resource type object instead?

hasanawad94 replied:
This is one is for setting the new field that we added https://github.com/shipwright-io/build/pull/1956 here

ayushsatyam146 reviewed (commented):
Thanks for all the changes @hasanawad94 . I am pointing out all the changes that I could spot below. Also please sign your commit after resolving for DCO

hasanawad94 commented:
/kind feature

adambkaplan commented:
nit - it looks like you are using merge commits to keep this branch up to date. This makes it harder to identify which changes are net-new code.

Please use `git rebase` instead.

SaschaSchwarze0 commented on docs/configuration.md:
```diff
@@ -37,6 +37,7 @@ The following environment variables are available:
 | `KUBE_API_BURST`                                 | Burst to use for the Kubernetes API client. See [Config.Burst]. A value of 0 or lower will use the default from client-go, which currently is 10. Default is 0.                                                                                                                                                                                            ...
 | `KUBE_API_QPS`                                   | QPS to use for the Kubernetes API client. See [Config.QPS]. A value of 0 or lower will use the default from client-go, which currently is 5. Default is 0.                                                                                                                                                                                                 ...
 | `VULNERABILITY_COUNT_LIMIT`                      | holds vulnerability count limit if vulnerability scan is enabled for the output image. If it is defined as 10, then it will output only 10 vulnerabilities sorted by severity in the buildrun status.Output. Default is 50.                                                                                                                                ...
+| `BUILDRUN_EXECUTOR`                      | Sets the kind of buildrun exectutor that will be used. Value can be `TaskRun` or `PipelineRun`. By default buildrun will use `TaskRun` for its build executor.                                                                                                                                                                                                     ...
```
```suggestion
| `BUILDRUN_EXECUTOR`                              | Sets the kind of buildrun exectutor that will be used. Value can be `TaskRun` or `PipelineRun`. By default buildrun will use `TaskRun` for its build executor.                                                                                                                                                                                                                                                                                                                                          |
```

SaschaSchwarze0 commented on pkg/reconciler/buildrun/buildrun.go:
```diff
@@ -320,46 +317,47 @@ func (r *ReconcileBuildRun) Reconcile(ctx context.Context, request reconcile.Req
-			ctxlog.Info(ctx, "updating BuildRun status with TaskRun name", namespace, request.Namespace, name, request.Name, "TaskRun", generatedTaskRun.Name)
+			// Set the deprecated TaskRunName field for backward compatibility
+			//nolint:staticcheck // Keep for backward compatibility for now
+			buildRun.Status.TaskRunName = &executorName
```
I suggest we only set it if it really is a TaskRun.

SaschaSchwarze0 commented on pkg/reconciler/buildrun/buildrun.go:
```diff
@@ -405,60 +403,60 @@ func (r *ReconcileBuildRun) Reconcile(ctx context.Context, request reconcile.Req
 				}
+				// Set the deprecated TaskRunName field for backward compatibility
+				//nolint:staticcheck // Keep for backward compatibility for now
+				buildRun.Status.TaskRunName = &executorName
```
I suggest we only set it if it really is a TaskRun.

SaschaSchwarze0 commented on pkg/reconciler/buildrun/controller.go:
```diff
@@ -77,18 +78,40 @@ func add(mgr manager.Manager, r reconcile.Reconciler, maxConcurrentReconciles in
-			// Process an update event when the old TR resource is not yet started and the new TR resource got a
-			// condition of the type Succeeded
-			if o.Status.StartTime.IsZero() && n.Status.GetCondition(apis.ConditionSucceeded) != nil {
+			// Check for completion time changes
+			if o.Status.CompletionTime == nil && n.Status.CompletionTime != nil {
 				return true
 			}
```
Why do we need to reconcile the BuildRun here again when the CompletionTime appeared ?

SaschaSchwarze0 commented on pkg/reconciler/buildrun/controller.go:
```diff
@@ -97,14 +120,62 @@ func add(mgr manager.Manager, r reconcile.Reconciler, maxConcurrentReconciles in
 		return err
 	}
 
+	// Watch for TaskRun events
```
Should not we look at the config and either watch TaskRuns or PipelineRuns ?

SaschaSchwarze0 commented on test/utils/v1beta1/controllers.go:
```diff
@@ -18,12 +19,25 @@ import (
 
 	c := buildconfig.NewDefaultConfig()
 
+	// Set executor if provided
```
Calling `SetConfigFromEnv` later changes again the BuildrunExecutor, does not it?

hasanawad94 commented:
Thanks for the review @SaschaSchwarze0. For the `GetPodName()` method its needed for the metrics collection.
I've removed that uncertain logic relying on the naming and moved it to fetch the pod name from the taskrun in the reconcile function incase of using `pipelinerun` https://github.com/shipwright-io/build/blob/6e897bffddbdf427e50be800c7df47ac2eb6275d/pkg/reconciler/buildrun/buildrun.go#L491-L498

adambkaplan commented on .github/workflows/ci.yml:
```diff
@@ -150,6 +150,9 @@ jobs:
           make test-integration
+      - name: Test-PipelineRun
+        run: |
+          BUILDRUN_EXECUTOR=PipelineRun ginkgo --focus-file="buildruns_to_pipelineruns_test.go" -v test/integration/...
```
nit (not blocking): ideally would like to have these be arguments to the make target, but I can live with this for now.

adambkaplan commented on pkg/config/config.go:
```diff
@@ -361,6 +364,11 @@ func (c *Config) SetConfigFromEnv() error {
+	// set environment variable for executor type
+	if executor := os.Getenv(controllerBuildrunExecutorEnvVar); executor != "" {
+		c.BuildrunExecutor = executor
+	}
```
nit (not blocking): ideally we should have some validation on this env var value and not accept it at face value.
